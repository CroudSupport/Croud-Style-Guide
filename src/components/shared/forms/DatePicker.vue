<template>
    <span>
        <div class="ui action input" :class="{fluid: fluid}">
            <input ref="pickerfield" @click.prevent.enter="updateDate" :value="display_date" type="text" :placeholder="placeholder">
            <button @click.prevent ref="pickerbutton" class="ui icon button">
                <i class="calendar outline icon"></i>
            </button>
        </div>
    </span>
</template>

<script>
    import _ from 'lodash'
    import Pikaday from 'pikaday'
    import croudDate from 'croud-date-parser'
    import moment from 'moment'

    export default {
        name: 'croud-date-picker',

        model: {
            prop: 'date',
        },

        props: {
            placeholder: {
                default: 'Date..',
            },

            date: {
                default: null,
            },

            min_date: {
                default: null,
            },

            max_date: {
                default: null,
            },

            reference: {
                default: null,
            },

            display: {
                default: 'D MMM YYYY',
            },

            fluid: {
                default: false,
            },

            settings: {
                type: Object,
                default() {
                    return {}
                },
            },
        },

        data() {
            return {
                internal_date: null,
                display_date: null,
                picker: null,
            }
        },

        watch: {
            date(value) {
                if (value && this.picker && this.internal_date && moment(value).format() !== this.internal_date.format()) {
                    this.picker.setMoment(moment(value))
                }
            },
        },

        methods: {
            updateDate() {
                this.internal_date = moment(croudDate(this.display_date))
            },

            create() {
                const inst = this
                if (this.picker) this.picker.destroy()
                const settings = {
                    format: this.display,
                    field: this.$refs.pickerfield,
                    trigger: this.$refs.pickerbutton,
                    onSelect: (val) => {
                        this.internal_date = moment(val)
                        this.display_date = this.internal_date.format(this.display)
                        this.$emit('date-selected', this.internal_date)
                        this.$emit('input', this.internal_date.format('YYYY-MM-DD'))
                        if (this.settings.afterSelect && typeof this.settings.afterSelect === 'function') { this.settings.afterSelect(this.internal_date) }
                    },
                }

                this.picker = new Pikaday(_.defaultsDeep(settings, this.settings))
                if (this.internal_date) this.picker.setMoment(this.internal_date)
            },
        },

        mounted() {
            if (this.date) {
                this.internal_date = moment(this.date)
                this.display_date = this.internal_date.format(this.display)
            }

            this.create()
        },

        beforeDestroy() {
            if (this.picker) this.picker.destroy()
        },
    }

</script>
