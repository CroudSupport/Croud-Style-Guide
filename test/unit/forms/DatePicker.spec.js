import Vue from 'vue'
import DatePicker from '../../../src/components/shared/forms/DatePicker'

const Constructor = Vue.extend(DatePicker)

const vm = new Constructor().$mount()

describe('Date Picker', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })
})
