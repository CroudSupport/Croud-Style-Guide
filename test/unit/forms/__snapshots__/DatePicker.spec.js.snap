// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Picker should match the snapshot 1`] = `
<span>
  <div
    class="ui action input"
  >
    <input
      type="text"
      placeholder="Date..."
    />
    <button
      class="ui icon button"
    >
      <i
        class="calendar outline icon"
      />
    </button>
  </div>
  <div
    class="pika-container"
    style="display: none;"
  >
    <div />
    <button
      class="ui basic button mini"
    >
      Today
    </button>
  </div>
</span>
`;
